drivers/adc/adc_ads114s08.c

```diff
@@ -0,0 +1,906 @@
+/*
+ * Copyright (c) 2023 SILA Embedded Solutions GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/drivers/spi.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/sys/util.h>
+
+#define ADC_CONTEXT_USES_KERNEL_TIMER 1
+#include "adc_context.h"
+
+#define DT_DRV_COMPAT ti_ads114s08
+
+LOG_MODULE_REGISTER(ads114s08, CONFIG_ADC_LOG_LEVEL);
+
+#define ADS114_CLK_FREQ_IN_MHZ				 4096
+#define ADS114_RESET_DELAY_TIME_IN_CLOCK_CYCLES		 4096
+#define ADS114_RESET_LOW_TIME_IN_CLOCK_CYCLES		 4
+#define ADS114_START_SYNC_PULSE_DURATION_IN_CLOCK_CYCLES 4
+#define ADS114_SETUP_TIME_IN_CLOCK_CYCLES		 32
+#define ADS114_INPUT_SELECTION_AINCOM			 12
+#define ADS114_RESOLUTION				 16
+#define ADS114_REF_INTERNAL				 2500
+
+#define ADS114_RESET_DELAY_TIME_IN_US                                                              \
+	(ADS114_RESET_DELAY_TIME_IN_CLOCK_CYCLES / ADS114_CLK_FREQ_IN_MHZ)
+#define ADS114_RESET_LOW_TIME_IN_US (ADS114_RESET_LOW_TIME_IN_CLOCK_CYCLES / ADS114_CLK_FREQ_IN_MHZ)
+#define ADS114_START_SYNC_PULSE_DURATION_IN_US                                                     \
+	(ADS114_START_SYNC_PULSE_DURATION_IN_CLOCK_CYCLES / ADS114_CLK_FREQ_IN_MHZ)
+#define ADS114_SETUP_TIME_IN_US (ADS114_SETUP_TIME_IN_CLOCK_CYCLES / ADS114_CLK_FREQ_IN_MHZ)
+
+enum ads114_command {
+	ADS114_COMMAND_NOP = 0x00,
+	ADS114_COMMAND_WAKEUP = 0x02,
+	ADS114_COMMAND_POWERDOWN = 0x04,
+	ADS114_COMMAND_RESET = 0x06,
+	ADS114_COMMAND_START = 0x08,
+	ADS114_COMMAND_STOP = 0x0A,
+	ADS114_COMMAND_SYOCAL = 0x16,
+	ADS114_COMMAND_SYGCAL = 0x17,
+	ADS114_COMMAND_SFOCAL = 0x19,
+	ADS114_COMMAND_RDATA = 0x12,
+	ADS114_COMMAND_RREG = 0x20,
+	ADS114_COMMAND_WREG = 0x40,
+};
+
+enum ads114_register {
+	ADS114_REGISTER_ID = 0x00,
+	ADS114_REGISTER_STATUS = 0x01,
+	ADS114_REGISTER_INPMUX = 0x02,
+	ADS114_REGISTER_PGA = 0x03,
+	ADS114_REGISTER_DATARATE = 0x04,
+	ADS114_REGISTER_REF = 0x05,
+	ADS114_REGISTER_IDACMAG = 0x06,
+	ADS114_REGISTER_IDACMUX = 0x07,
+	ADS114_REGISTER_VBIAS = 0x08,
+	ADS114_REGISTER_SYS = 0x09,
+	ADS114_REGISTER_OFCAL0 = 0x0B,
+	ADS114_REGISTER_OFCAL1 = 0x0C,
+	ADS114_REGISTER_FSCAL0 = 0x0E,
+	ADS114_REGISTER_FSCAL1 = 0x0F,
+	ADS114_REGISTER_GPIODAT = 0x10,
+	ADS114_REGISTER_GPIOCON = 0x11,
+};
+
+#define ADS114_REGISTER_GENERATE_BIT_MASK(length) ((1 << length) - 1)
+#define ADS114_REGISTER_GET_VALUE(value, pos, length)                                              \
+	((value >> pos) & ADS114_REGISTER_GENERATE_BIT_MASK(length))
+#define ADS114_REGISTER_SET_VALUE(target, value, pos, length)                                      \
+	target &= ~(ADS114_REGISTER_GENERATE_BIT_MASK(length) << pos);                             \
+	target |= ((value & ADS114_REGISTER_GENERATE_BIT_MASK(length)) << pos)
```

Comment:
Please examine if the macros in `include/zephyr/sys/util.h` can be used here (`FIELD_PREP()`, `FIELD_GET()`, `GENMASK()`, ...)

Reply:
done